tests/infra_test.go

```diff
@@ -696,6 +697,9 @@ var _ = Describe("[Serial]Infrastructure", func() {
 			for ix := 0; ix < concurrency; ix++ {
 				err := <-errors
 				if err != nil {
+					urlErr, ok := err.(*netURL.Error)
+					Expect(ok).To(BeTrue())
+					Expect(urlErr.Err.Error()).To(ContainSubstring("Client.Timeout exceeded while awaiting headers"))
```

Comment:
does this error string belong to another library?
if it's internal then it's ok, but if it's not then this string can change in the next version update, and the test will break.

Reply:
its internal, it a known error, that is used for long time in network libraries.

Reply:
Where is this error thrown? I do not seem to find it (I grep for it and failed).

Also, the previous assertion will not give any valuable information if it fails, and it is pretty hard to follow the logic here.

- I suggest you try to find the relevant erros using the new error functionalities in Go 1.13 [1].
  (see `Examining errors with Is and As` section)
- I suggest with the previous point to make it read something like: If this is not the expected error, break from the loop. If when finishing the loop, assert against the error such that we will get the reasoning of the failure.
(you may need a helper function to return the error is the loop/retry fails eventually)

1. https://blog.golang.org/go1.13-errors

Reply:
> Where is this error thrown? I do not seem to find it (I grep for it and failed).
>
Print the error and see it returns a (*url.Error)
```
client: request: &{GET https://10.244.140.74:8443/metrics HTTP/1.1 1 1 map[] <nil> <nil> 0 [] false 10.244.140.74:8443 map[] map[] <nil> map[]   <nil> <nil> <nil> 0xc0000de010} #0: Get https://10.244.140.74:8443/metrics: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```
> Also, the previous assertion will not give any valuable information if it fails, and it is pretty hard to follow the logic here.
> 
> * I suggest you try to find the relevant erros using the new error functionalities in Go 1.13 [1].
>   (see `Examining errors with Is and As` section)
> * I suggest with the previous point to make it read something like: If this is not the expected error, break from the loop. If when finishing the loop, assert against the error such that we will get the reasoning of the failure.
>   (you may need a helper function to return the error is the loop/retry fails eventually)

We already getting the reasoning on the error upon the failed Expect, i find the logic above not suitable and much harder to follow.
I tried to make it simpler, and this is the simplest and trustful we found (see discussions on 4140 if you wish also please, since this PR is part of that one).
Also please search `IsProbableEOF` vendor/k8s.io/apimachinery/pkg/util/net/http.go
The same method is used there.


> 
> 1. https://blog.golang.org/go1.13-errors

Reply:
> > Where is this error thrown? I do not seem to find it (I grep for it and failed).
> 
> Print the error and see it returns a (*url.Error), and this is the string of the error which is ok, its the ip stack of some compiled lib that returns it i believe.

This contradicts the answer you gave @RamLavi when you said "its internal".
If it is not internal his comment stands and indeed it is problematic to depend on the string.

> We already getting the reasoning on the error upon the failed Expect, i find the logic above not suitable and much harder to follow.

If the error is not of the right type, no info will be provided.

I find the existing code exposing too much details which I need to decode in order to understand what is done here.
I want to read the logic without being ditracted by details, this way I will understand faster and it will help future maintenance of this code.

> 
> Also the only expected error is url.Error, so we can use Expect and fail the test instead capturing, breaking, rechecking. This is a reason for example why not to use the `As` with the logic of breaking and so on, but to keep it simple.

If it is the wrong error, you are expected to let it show at the test failure and not to silently swallow it.
`As` performs the type check for you, without the need to "manually" do it through "casting"/type assertion.

> I tried to make it simpler, and this is the simplest and trustful we found (see discussions on 4140 if you wish also please, since this PR is part of that one).
> Also please see `IsProbableEOF` vendor/k8s.io/apimachinery/pkg/util/net/http.go
> The same method is used there.

You have been suggested a different way, something more "modern".
BTW, that ref is a function that does on thing (returning if this is a specific error or not), unlike the embedded version in this PR.

Reply:
```
    Expected
        <string>: dial tcp [fd10:244::8c48]:8443: connect: network is unreachable
    to contain substring
        <string>: Client.Timeout exceeded while awaiting headers
```
This what i meant that in case the string is something else we will see it, for example since there is a bug with ipv6 we see the first error happens instead the expect which is the 2nd line.

There are too much details in this comment, i honestly find it over-engineering,
i will try to first start with a function pretty similar to `IsProbableEOF` but with `As` and lets iterate from there please.

Reply:
The lack of information is at this assertion: `Expect(ok).To(BeTrue())`.

Asserting on a specific string from an external library is incorrect as commented on by the first comment.
I do not know what is over-engineered in the comments, they point on specific problem of maintenance to this piece of code.

Reply:
tried to take care about it, PTAL